File: modules/expo-vault/android/src/main/java/expo/modules/vault/VaultCrypto.kt
package expo.modules.vault

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import androidx.exifinterface.media.ExifInterface
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.nio.ByteBuffer
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * Streaming AES-256-GCM container shared with the iOS implementation.
 *
 * Layout: magic "SVC1", chunk size (Int32 BE), then per chunk an Int32 BE
 * length followed by nonce (12) + ciphertext + tag (16). The chunk index is
 * bound as additional authenticated data so chunks cannot be reordered or
 * dropped. Files written by the 2025 prototype have no header and keep their
 * IV in SharedPreferences; [legacyIv] handles those.
 */
object VaultCrypto {
    private val MAGIC = "SVC1".toByteArray(Charsets.US_ASCII)
    const val CHUNK_SIZE = 1_048_576
    private const val NONCE_SIZE = 12
    private const val TAG_BITS = 128

    fun encryptFile(source: File, destination: File, key: SecretKey) {
        FileInputStream(source).use { input ->
            DataOutputStream(FileOutputStream(destination).buffered()).use { output ->
                output.write(MAGIC)
                output.writeInt(CHUNK_SIZE)
                val buffer = ByteArray(CHUNK_SIZE)
                var index = 0
                while (true) {
                    val read = readFully(input, buffer)
                    if (read <= 0) break
                    // Keystore keys generate their own random IV per init.
                    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                    cipher.init(Cipher.ENCRYPT_MODE, key)
                    cipher.updateAAD(be32(index))
                    val sealed = cipher.doFinal(buffer, 0, read)
                    val iv = cipher.iv
                    output.writeInt(iv.size + sealed.size)
                    output.write(iv)
                    output.write(sealed)
                    index += 1
                }
            }
        }
    }

    fun encryptBytes(bytes: ByteArray, destination: File, key: SecretKey) {
        val temp = File.createTempFile("vault", null, destination.parentFile)
        try {
            temp.writeBytes(bytes)
            encryptFile(temp, destination, key)
        } finally {
            temp.delete()
        }
    }

    /** Returns true when the file uses the chunked container. */
    fun isChunked(file: File): Boolean {
        if (file.length() < MAGIC.size) return false
        FileInputStream(file).use { input ->
            val header = ByteArray(MAGIC.size)
            val read = readFully(input, header)
            return read == MAGIC.size && header.contentEquals(MAGIC)
        }
    }

    fun decryptFile(source: File, destination: File, key: SecretKey, legacyIv: ByteArray?) {
        if (!isChunked(source)) {
            val iv = legacyIv ?: throw IllegalStateException("Legacy file without IV")
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(TAG_BITS, iv))
            destination.parentFile?.mkdirs()
            destination.writeBytes(cipher.doFinal(source.readBytes()))
            return
        }
        destination.parentFile?.mkdirs()
        DataInputStream(FileInputStream(source).buffered()).use { input ->
            input.skipBytes(MAGIC.size)
            input.readInt() // chunk size, informational
            FileOutputStream(destination).buffered().use { output ->
                var index = 0
                while (input.available() > 0) {
                    val length = input.readInt()
                    val iv = ByteArray(NONCE_SIZE)
                    input.readFully(iv)
                    val sealed = ByteArray(length - NONCE_SIZE)
                    input.readFully(sealed)
                    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                    cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(TAG_BITS, iv))
                    cipher.updateAAD(be32(index))
                    output.write(cipher.doFinal(sealed))
                    index += 1
                }
            }
        }
    }

    /** Downsampled, orientation-corrected JPEG for list thumbnails. */
    fun thumbnailJpeg(source: File, maxPixelSize: Int): ByteArray {
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(source.path, bounds)
        var sample = 1
        while (bounds.outWidth / (sample * 2) >= maxPixelSize || bounds.outHeight / (sample * 2) >= maxPixelSize) {
            sample *= 2
        }
        val options = BitmapFactory.Options().apply { inSampleSize = sample }
        val decoded = BitmapFactory.decodeFile(source.path, options)
            ?: throw IllegalStateException("Could not decode image")
        val scale = minOf(1f, maxPixelSize.toFloat() / maxOf(decoded.width, decoded.height))
        val rotation = when (ExifInterface(source.path).getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)) {
            ExifInterface.ORIENTATION_ROTATE_90 -> 90f
            ExifInterface.ORIENTATION_ROTATE_180 -> 180f
            ExifInterface.ORIENTATION_ROTATE_270 -> 270f
            else -> 0f
        }
        val matrix = Matrix().apply {
            postScale(scale, scale)
            if (rotation != 0f) postRotate(rotation)
        }
        val bitmap = Bitmap.createBitmap(decoded, 0, 0, decoded.width, decoded.height, matrix, true)
        val output = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 80, output)
        if (bitmap !== decoded) decoded.recycle()
        bitmap.recycle()
        return output.toByteArray()
    }

    private fun readFully(input: FileInputStream, buffer: ByteArray): Int {
        var total = 0
        while (total < buffer.size) {
            val read = input.read(buffer, total, buffer.size - total)
            if (read < 0) break
            total += read
        }
        return total
    }

    private fun be32(value: Int): ByteArray = ByteBuffer.allocate(4).putInt(value).array()
}

File: modules/expo-vault/ios/VaultCrypto.swift
import CryptoKit
import Foundation
import ImageIO
import UIKit

/// Streaming AES-256-GCM container shared with the Android implementation.
///
/// Layout: magic "SVC1", chunk size (UInt32 BE), then per chunk a UInt32 BE
/// length followed by nonce (12) + ciphertext + tag (16). The chunk index is
/// bound as additional authenticated data so chunks cannot be reordered or
/// dropped. Files written by the 2025 prototype are a single CryptoKit sealed
/// box with no header and are still readable.
enum VaultCrypto {
  static let magic = Data("SVC1".utf8)
  static let chunkSize = 1_048_576
  static let nonceSize = 12
  static let tagSize = 16

  enum CryptoError: Error, LocalizedError {
    case corrupt(String)
    var errorDescription: String? {
      if case let .corrupt(reason) = self { return "Vault file is corrupt: \(reason)" }
      return nil
    }
  }

  static func encryptFile(from source: URL, to destination: URL, key: SymmetricKey) throws {
    let input = try FileHandle(forReadingFrom: source)
    defer { try? input.close() }
    FileManager.default.createFile(atPath: destination.path, contents: nil)
    let output = try FileHandle(forWritingTo: destination)
    defer { try? output.close() }

    output.write(magic)
    output.write(be32(UInt32(chunkSize)))
    var index: UInt32 = 0
    while true {
      let chunk = input.readData(ofLength: chunkSize)
      if chunk.isEmpty { break }
      let sealed = try AES.GCM.seal(chunk, using: key, authenticating: be32(index))
      guard let combined = sealed.combined else { throw CryptoError.corrupt("seal") }
      output.write(be32(UInt32(combined.count)))
      output.write(combined)
      index += 1
    }
  }

  static func decryptFile(from source: URL, to destination: URL, key: SymmetricKey) throws {
    let input = try FileHandle(forReadingFrom: source)
    defer { try? input.close() }
    let header = input.readData(ofLength: magic.count)
    guard header == magic else {
      // Legacy single sealed box.
      try? input.close()
      let sealedData = try Data(contentsOf: source)
      let box = try AES.GCM.SealedBox(combined: sealedData)
      let plain = try AES.GCM.open(box, using: key)
      try plain.write(to: destination, options: .atomic)
      return
    }
    _ = input.readData(ofLength: 4) // chunk size, informational

    try FileManager.default.createDirectory(at: destination.deletingLastPathComponent(), withIntermediateDirectories: true)
    FileManager.default.createFile(atPath: destination.path, contents: nil)
    let output = try FileHandle(forWritingTo: destination)
    defer { try? output.close() }

    var index: UInt32 = 0
    while true {
      let lengthData = input.readData(ofLength: 4)
      if lengthData.isEmpty { break }
      guard lengthData.count == 4 else { throw CryptoError.corrupt("length") }
      let length = Int(readBe32(lengthData))
      let combined = input.readData(ofLength: length)
      guard combined.count == length else { throw CryptoError.corrupt("chunk \(index)") }
      let box = try AES.GCM.SealedBox(combined: combined)
      let plain = try AES.GCM.open(box, using: key, authenticating: be32(index))
      output.write(plain)
      index += 1
    }
  }

  static func encryptData(_ data: Data, to destination: URL, key: SymmetricKey) throws {
    let temp = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
    try data.write(to: temp, options: .atomic)
    defer { try? FileManager.default.removeItem(at: temp) }
    try encryptFile(from: temp, to: destination, key: key)
  }

  /// Downsampled JPEG for list thumbnails; orientation is applied so the
  /// thumbnail matches the photo the user took.
  static func thumbnailJPEG(from source: URL, maxPixelSize: Int) throws -> Data {
    let options: [CFString: Any] = [
      kCGImageSourceCreateThumbnailFromImageAlways: true,
      kCGImageSourceCreateThumbnailWithTransform: true,
      kCGImageSourceThumbnailMaxPixelSize: maxPixelSize,
      kCGImageSourceShouldCache: false,
    ]
    guard let imageSource = CGImageSourceCreateWithURL(source as CFURL, nil),
      let cgImage = CGImageSourceCreateThumbnailAtIndex(imageSource, 0, options as CFDictionary),
      let jpeg = UIImage(cgImage: cgImage).jpegData(compressionQuality: 0.8)
    else {
      throw CryptoError.corrupt("thumbnail")
    }
    return jpeg
  }

  private static func be32(_ value: UInt32) -> Data {
    var big = value.bigEndian
    return Data(bytes: &big, count: 4)
  }

  private static func readBe32(_ data: Data) -> UInt32 {
    data.withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
  }
}
